# MOVIE REVIEWS

We have a Dataset containing different movie reviews.  
We have to predict whether a movie review is positive or negative.

## Import Dataset

In [1]:
from nltk.corpus import movie_reviews

In [2]:
import nltk
# nltk.download('movie_reviews')

In [3]:
movie_reviews.categories()  # movie reviews has 2 categories -> negative and positive

['neg', 'pos']

In [4]:
movie_reviews.fileids()  # list of files corresponding to 2 different categories

['neg/cv000_29416.txt',
 'neg/cv001_19502.txt',
 'neg/cv002_17424.txt',
 'neg/cv003_12683.txt',
 'neg/cv004_12641.txt',
 'neg/cv005_29357.txt',
 'neg/cv006_17022.txt',
 'neg/cv007_4992.txt',
 'neg/cv008_29326.txt',
 'neg/cv009_29417.txt',
 'neg/cv010_29063.txt',
 'neg/cv011_13044.txt',
 'neg/cv012_29411.txt',
 'neg/cv013_10494.txt',
 'neg/cv014_15600.txt',
 'neg/cv015_29356.txt',
 'neg/cv016_4348.txt',
 'neg/cv017_23487.txt',
 'neg/cv018_21672.txt',
 'neg/cv019_16117.txt',
 'neg/cv020_9234.txt',
 'neg/cv021_17313.txt',
 'neg/cv022_14227.txt',
 'neg/cv023_13847.txt',
 'neg/cv024_7033.txt',
 'neg/cv025_29825.txt',
 'neg/cv026_29229.txt',
 'neg/cv027_26270.txt',
 'neg/cv028_26964.txt',
 'neg/cv029_19943.txt',
 'neg/cv030_22893.txt',
 'neg/cv031_19540.txt',
 'neg/cv032_23718.txt',
 'neg/cv033_25680.txt',
 'neg/cv034_29446.txt',
 'neg/cv035_3343.txt',
 'neg/cv036_18385.txt',
 'neg/cv037_19798.txt',
 'neg/cv038_9781.txt',
 'neg/cv039_5963.txt',
 'neg/cv040_8829.txt',
 'neg/cv041_22364.txt',


In [5]:
len(movie_reviews.fileids())

2000

-> 2000 movie_reviews, 1000 +ve and 1000 -ve

In [6]:
movie_reviews.fileids('neg')  #only -ve reviews

['neg/cv000_29416.txt',
 'neg/cv001_19502.txt',
 'neg/cv002_17424.txt',
 'neg/cv003_12683.txt',
 'neg/cv004_12641.txt',
 'neg/cv005_29357.txt',
 'neg/cv006_17022.txt',
 'neg/cv007_4992.txt',
 'neg/cv008_29326.txt',
 'neg/cv009_29417.txt',
 'neg/cv010_29063.txt',
 'neg/cv011_13044.txt',
 'neg/cv012_29411.txt',
 'neg/cv013_10494.txt',
 'neg/cv014_15600.txt',
 'neg/cv015_29356.txt',
 'neg/cv016_4348.txt',
 'neg/cv017_23487.txt',
 'neg/cv018_21672.txt',
 'neg/cv019_16117.txt',
 'neg/cv020_9234.txt',
 'neg/cv021_17313.txt',
 'neg/cv022_14227.txt',
 'neg/cv023_13847.txt',
 'neg/cv024_7033.txt',
 'neg/cv025_29825.txt',
 'neg/cv026_29229.txt',
 'neg/cv027_26270.txt',
 'neg/cv028_26964.txt',
 'neg/cv029_19943.txt',
 'neg/cv030_22893.txt',
 'neg/cv031_19540.txt',
 'neg/cv032_23718.txt',
 'neg/cv033_25680.txt',
 'neg/cv034_29446.txt',
 'neg/cv035_3343.txt',
 'neg/cv036_18385.txt',
 'neg/cv037_19798.txt',
 'neg/cv038_9781.txt',
 'neg/cv039_5963.txt',
 'neg/cv040_8829.txt',
 'neg/cv041_22364.txt',


In [7]:
movie_reviews.fileids('pos')  #only +ve reviews

['pos/cv000_29590.txt',
 'pos/cv001_18431.txt',
 'pos/cv002_15918.txt',
 'pos/cv003_11664.txt',
 'pos/cv004_11636.txt',
 'pos/cv005_29443.txt',
 'pos/cv006_15448.txt',
 'pos/cv007_4968.txt',
 'pos/cv008_29435.txt',
 'pos/cv009_29592.txt',
 'pos/cv010_29198.txt',
 'pos/cv011_12166.txt',
 'pos/cv012_29576.txt',
 'pos/cv013_10159.txt',
 'pos/cv014_13924.txt',
 'pos/cv015_29439.txt',
 'pos/cv016_4659.txt',
 'pos/cv017_22464.txt',
 'pos/cv018_20137.txt',
 'pos/cv019_14482.txt',
 'pos/cv020_8825.txt',
 'pos/cv021_15838.txt',
 'pos/cv022_12864.txt',
 'pos/cv023_12672.txt',
 'pos/cv024_6778.txt',
 'pos/cv025_3108.txt',
 'pos/cv026_29325.txt',
 'pos/cv027_25219.txt',
 'pos/cv028_26746.txt',
 'pos/cv029_18643.txt',
 'pos/cv030_21593.txt',
 'pos/cv031_18452.txt',
 'pos/cv032_22550.txt',
 'pos/cv033_24444.txt',
 'pos/cv034_29647.txt',
 'pos/cv035_3954.txt',
 'pos/cv036_16831.txt',
 'pos/cv037_18510.txt',
 'pos/cv038_9749.txt',
 'pos/cv039_6170.txt',
 'pos/cv040_8276.txt',
 'pos/cv041_21113.txt',
 

In [8]:
# to access words written in reviews

movie_reviews.words(movie_reviews.fileids()[5])  # accessing 6th review (indexing from 0)

['capsule', ':', 'in', '2176', 'on', 'the', 'planet', ...]

In [9]:
# Now, creating a 2000 size array or list, containing tuples of list of words in document and its category 
# [([word1.1, word1.2,...], category1), ([word2.1, word2.2,...], category2), ...]

documents = []  # array or list
for category in movie_reviews.categories():  # going through each category one by one
    for fileid in movie_reviews.fileids(category):  # first get +ve or -ve (any one)
        documents.append((movie_reviews.words(fileid), category))  # appending words and its category that we get 
                                                                   # from these particular file id in tuple format 
documents[0:5]

[(['plot', ':', 'two', 'teen', 'couples', 'go', 'to', ...], 'neg'),
 (['the', 'happy', 'bastard', "'", 's', 'quick', 'movie', ...], 'neg'),
 (['it', 'is', 'movies', 'like', 'these', 'that', 'make', ...], 'neg'),
 (['"', 'quest', 'for', 'camelot', '"', 'is', 'warner', ...], 'neg'),
 (['synopsis', ':', 'a', 'mentally', 'unstable', 'man', ...], 'neg')]

In [11]:
# shuffling the negative and positive reviews so that splitting can be done

import random
random.shuffle(documents)
documents[0:5]

[(['the', 'premise', 'of', 'turbulence', 'is', 'i', "'", ...], 'neg'),
 (['the', 'full', 'monty', 'is', 'a', 'whole', 'lot', ...], 'pos'),
 (['review', ':', 'ghost', 'dog', ':', 'the', 'way', ...], 'pos'),
 (['"', 'tina', '!', '!', '!', 'fetch', 'me', 'the', ...], 'neg'),
 (['warning', '!', ':', 'may', 'contain', 'some', 'mild', ...], 'pos')]

## Data Cleaning

-> removing stop words, punctuations.  
-> apply lemmatization

In [12]:
# Lemmatize

from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()  # creating object

In [13]:
from nltk.corpus import wordnet

def get_simple_pos(tag):
    if tag.startswith('J'):
        return wordnet.ADJ
    elif tag.startswith('N'):
        return wordnet.NOUN
    elif tag.startswith('V'):
        return wordnet.VERB
    elif tag.startswith('R'):
        return wordnet.ADV
    else:
        return wordnet.NOUN

In [14]:
# Creating list of stop words and adding punctuations also

from nltk.corpus import stopwords
import string

stops = set(stopwords.words("english"))
punctuations = list(string.punctuation)
stops.update(punctuations)  # updating list with punctuations
stops

{'!',
 '"',
 '#',
 '$',
 '%',
 '&',
 "'",
 '(',
 ')',
 '*',
 '+',
 ',',
 '-',
 '.',
 '/',
 ':',
 ';',
 '<',
 '=',
 '>',
 '?',
 '@',
 '[',
 '\\',
 ']',
 '^',
 '_',
 '`',
 'a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 'her',
 'here',
 'hers',
 'herself',
 'him',
 'himself',
 'his',
 'how',
 'i',
 'if',
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it's",
 'its',
 'itself',
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'need

In [15]:
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [18]:
def clean_review(words):
    output_words = []
    for w in words:
        if w.lower() not in stops:  
            pos = pos_tag([w]) # passing array
            clean_word = lemmatizer.lemmatize(w, pos = get_simple_pos(pos[0][1])) 
            output_words.append(clean_word.lower()) 
    return output_words

In [19]:
# Explaination:-

def clean_review(words):
    output_words = []
    for w in words:
        if w.lower() not in stops:  # stops -> list of stop words and punctuations # lowering the words first
            pos = pos_tag([w])  # finding pos_tag (part of speech) of w, not of lower w, bcoz we don't want to loose 
                              #information
            clean_word = lemmatizer.lemmatize(w, pos = get_simple_pos(pos[0][1])) # we have to change postags from NN 
                                                                                    # to'n'(example) to be used in 
                                                                                    # lammatize function
                        #   pos[0][1] -> 1st index of tuple at 0th index of array w
                        #   Ex: pos[0][1] of [([word1.1,..], category1), ([word2.1,..], category2), ...]  = category1
            output_words.append(clean_word.lower())  # appending after lowering word
    return output_words

In [20]:
# Going through each document and cleaning words

documents = [(clean_review(document), category) for document, category in documents]

In [21]:
# Explaination:-

documents = [(clean_review(x), category) for x, category in documents]
#    (clean_review(x), category) -> adding a new tuple           x, category -> accessing a tuple

In [22]:
documents[0]

(['premise',
  'turbulence',
  'sure',
  'familiar',
  'u',
  'see',
  'passenger',
  '57',
  'executive',
  'decision',
  'countless',
  'flick',
  'good',
  'way',
  'know',
  'terrorist',
  'take',
  'plane',
  'ask',
  'police',
  'ground',
  'turbulence',
  'start',
  'convict',
  'ray',
  'liotta',
  'accuse',
  'crime',
  'think',
  'way',
  'could',
  'open',
  'kind',
  'take',
  'subsequently',
  'transport',
  'commerical',
  'plane',
  'la',
  'already',
  'problem',
  'way',
  'convict',
  'would',
  'transport',
  'commercial',
  'flight',
  'normal',
  'passangers',
  'would',
  'anyway',
  'christmas',
  'eve',
  'everyone',
  'wait',
  'get',
  'la',
  'spend',
  'christmas',
  'eve',
  'well',
  'convict',
  'one',
  'manages',
  'stop',
  'flight',
  'proverbial',
  'track',
  'liotta',
  'still',
  'calm',
  'ever',
  'take',
  'convict',
  'also',
  'eye',
  'lauren',
  'holly',
  'flight',
  'attendent',
  'seem',
  'grow',
  'closer',
  'plot',
  'stop',
  'like'

## Building Feature Set

**NLTK classifiers require data to be in this format:  
Array of Tuples, where each Tuple has Dictionary of features and category.**

**Splitting training and testing data**

In [26]:
training_documents = documents[0:1500]
testing_documents = documents[1500:]

In [29]:
# Creating array containing all the words

all_words = []
for doc in training_documents:
    all_words += doc[0]  #training-documents is an array containing tuples of (word, category)
                         # doc[0] -> words, doc[1] -> category
# all_words

In [30]:
# Finding frequency distribution

import nltk

In [31]:
freq = nltk.FreqDist(all_words)
freq

FreqDist({'film': 8302, 'movie': 5242, 'one': 4509, 'make': 3177, 'like': 2948, 'character': 2896, 'get': 2744, 'see': 2374, 'go': 2338, 'time': 2187, ...})

In [32]:
freq.most_common(15)  # first 15 most frequent words

[('film', 8302),
 ('movie', 5242),
 ('one', 4509),
 ('make', 3177),
 ('like', 2948),
 ('character', 2896),
 ('get', 2744),
 ('see', 2374),
 ('go', 2338),
 ('time', 2187),
 ('well', 2112),
 ('scene', 1970),
 ('even', 1963),
 ('story', 1806),
 ('good', 1785)]

In [33]:
freq.most_common(3000)  # an array of tuples

[('film', 8302),
 ('movie', 5242),
 ('one', 4509),
 ('make', 3177),
 ('like', 2948),
 ('character', 2896),
 ('get', 2744),
 ('see', 2374),
 ('go', 2338),
 ('time', 2187),
 ('well', 2112),
 ('scene', 1970),
 ('even', 1963),
 ('story', 1806),
 ('good', 1785),
 ('take', 1649),
 ('would', 1619),
 ('much', 1516),
 ('come', 1499),
 ('also', 1472),
 ('two', 1456),
 ('life', 1450),
 ('way', 1439),
 ('know', 1413),
 ('bad', 1413),
 ('end', 1409),
 ('give', 1405),
 ('first', 1374),
 ('seem', 1371),
 ('look', 1364),
 ('--', 1347),
 ('work', 1340),
 ('year', 1303),
 ('thing', 1242),
 ('say', 1196),
 ('plot', 1195),
 ('play', 1169),
 ('really', 1154),
 ('little', 1142),
 ('people', 1124),
 ('show', 1074),
 ('man', 1061),
 ('could', 1056),
 ('u', 1049),
 ('never', 1036),
 ('love', 1035),
 ('star', 1026),
 ('great', 1019),
 ('best', 1010),
 ('try', 1003),
 ('director', 996),
 ('performance', 977),
 ('new', 977),
 ('many', 958),
 ('want', 924),
 ('big', 915),
 ('watch', 900),
 ('action', 895),
 ('find

**Features:-**

In [34]:
common = freq.most_common(3000)
features = [i[0] for i in common]  # features contains only words and not frequencies
                                   # Here, features is a list or array containing 3000 words or features
features

['film',
 'movie',
 'one',
 'make',
 'like',
 'character',
 'get',
 'see',
 'go',
 'time',
 'well',
 'scene',
 'even',
 'story',
 'good',
 'take',
 'would',
 'much',
 'come',
 'also',
 'two',
 'life',
 'way',
 'know',
 'bad',
 'end',
 'give',
 'first',
 'seem',
 'look',
 '--',
 'work',
 'year',
 'thing',
 'say',
 'plot',
 'play',
 'really',
 'little',
 'people',
 'show',
 'man',
 'could',
 'u',
 'never',
 'love',
 'star',
 'great',
 'best',
 'try',
 'director',
 'performance',
 'new',
 'many',
 'want',
 'big',
 'watch',
 'action',
 'find',
 'actor',
 'think',
 'role',
 'act',
 'another',
 'world',
 'back',
 'turn',
 'still',
 'audience',
 'something',
 'day',
 'feel',
 'however',
 'set',
 'comedy',
 'old',
 'though',
 'use',
 'guy',
 'begin',
 'part',
 'point',
 'every',
 'real',
 'around',
 'fact',
 'enough',
 'interest',
 'may',
 'write',
 'cast',
 'run',
 'last',
 'funny',
 'young',
 'actually',
 'name',
 'script',
 'right',
 'effect',
 'woman',
 'long',
 'nothing',
 'minute',
 'alt

Now, we have to create a dictionary for each document, stating True or False as value -> wheather each feature comes in each document or not  
Keys - Features  
Values - True or False

In [35]:
training_documents[0]  # tuple of list of words and the category of document. 
                       # Ex: ([word1, word2, word3,...], category of doc.)

(['premise',
  'turbulence',
  'sure',
  'familiar',
  'u',
  'see',
  'passenger',
  '57',
  'executive',
  'decision',
  'countless',
  'flick',
  'good',
  'way',
  'know',
  'terrorist',
  'take',
  'plane',
  'ask',
  'police',
  'ground',
  'turbulence',
  'start',
  'convict',
  'ray',
  'liotta',
  'accuse',
  'crime',
  'think',
  'way',
  'could',
  'open',
  'kind',
  'take',
  'subsequently',
  'transport',
  'commerical',
  'plane',
  'la',
  'already',
  'problem',
  'way',
  'convict',
  'would',
  'transport',
  'commercial',
  'flight',
  'normal',
  'passangers',
  'would',
  'anyway',
  'christmas',
  'eve',
  'everyone',
  'wait',
  'get',
  'la',
  'spend',
  'christmas',
  'eve',
  'well',
  'convict',
  'one',
  'manages',
  'stop',
  'flight',
  'proverbial',
  'track',
  'liotta',
  'still',
  'calm',
  'ever',
  'take',
  'convict',
  'also',
  'eye',
  'lauren',
  'holly',
  'flight',
  'attendent',
  'seem',
  'grow',
  'closer',
  'plot',
  'stop',
  'like'

In [36]:
# Creating Dictionary

def get_feature_dictionary(words):  # words -> list of words in tuple ([word1, word2, ...], category)
    current_features = {}
    words_set = set(words)
    for w in features:
        current_features[w] = w in words_set  # w -> key,  value -> True - if w is in document
                                              #                     False - if w not in document
    return current_features

In [37]:
get_feature_dictionary(training_documents[0][0])  

{'film': False,
 'movie': True,
 'one': True,
 'make': False,
 'like': True,
 'character': False,
 'get': True,
 'see': True,
 'go': False,
 'time': False,
 'well': True,
 'scene': False,
 'even': True,
 'story': False,
 'good': True,
 'take': True,
 'would': True,
 'much': False,
 'come': False,
 'also': True,
 'two': False,
 'life': False,
 'way': True,
 'know': True,
 'bad': False,
 'end': False,
 'give': True,
 'first': False,
 'seem': True,
 'look': True,
 '--': False,
 'work': False,
 'year': False,
 'thing': False,
 'say': False,
 'plot': True,
 'play': False,
 'really': False,
 'little': False,
 'people': False,
 'show': False,
 'man': False,
 'could': True,
 'u': True,
 'never': False,
 'love': False,
 'star': True,
 'great': False,
 'best': False,
 'try': False,
 'director': False,
 'performance': False,
 'new': False,
 'many': False,
 'want': False,
 'big': False,
 'watch': False,
 'action': False,
 'find': False,
 'actor': False,
 'think': True,
 'role': False,
 'act': Fals

In [38]:
training_data = [(get_feature_dictionary(doc), category)for doc, category in training_documents]
testing_data = [(get_feature_dictionary(doc), category)for doc, category in testing_documents]

In [39]:
training_data[0]

({'film': False,
  'movie': True,
  'one': True,
  'make': False,
  'like': True,
  'character': False,
  'get': True,
  'see': True,
  'go': False,
  'time': False,
  'well': True,
  'scene': False,
  'even': True,
  'story': False,
  'good': True,
  'take': True,
  'would': True,
  'much': False,
  'come': False,
  'also': True,
  'two': False,
  'life': False,
  'way': True,
  'know': True,
  'bad': False,
  'end': False,
  'give': True,
  'first': False,
  'seem': True,
  'look': True,
  '--': False,
  'work': False,
  'year': False,
  'thing': False,
  'say': False,
  'plot': True,
  'play': False,
  'really': False,
  'little': False,
  'people': False,
  'show': False,
  'man': False,
  'could': True,
  'u': True,
  'never': False,
  'love': False,
  'star': True,
  'great': False,
  'best': False,
  'try': False,
  'director': False,
  'performance': False,
  'new': False,
  'many': False,
  'want': False,
  'big': False,
  'watch': False,
  'action': False,
  'find': False,
  

### using Naive Bayes Classifier

Classifier needs data to be in the format:-  
list or array of tuples, where each tuple contains dictionary of features and category of document, each dictionary conatins boolean value of each feature, whether that particular feature is present in our document or not.  

Syntax:-  

**[({feature1.1:True/False, feature1.2:True/False}, category1), ({{feature2.1:True/False, feature2.2:True/False}}, category2), ({{feature3.1:True/False, feature3.2:True/False}}, category3), ...]**

In [40]:
from nltk import NaiveBayesClassifier

In [41]:
classifier = NaiveBayesClassifier.train(training_data)  # training classifier

In [43]:
# nltk has classify module having accuracy() function

nltk.classify.accuracy(classifier, testing_data)  # finding accuracy of this classifier on the testing data

0.776

-> we've got an accuracy of 0.776 or 77.6%

In [46]:
classifier.show_most_informative_features(15)  # show_most_informative_features() -> tells us the most informative 
                                               # feature this classifier has


Most Informative Features
             outstanding = True              pos : neg    =     12.6 : 1.0
               ludicrous = True              neg : pos    =     12.1 : 1.0
                   damon = True              pos : neg    =      9.6 : 1.0
             wonderfully = True              pos : neg    =      8.9 : 1.0
            breathtaking = True              pos : neg    =      8.5 : 1.0
                 refresh = True              pos : neg    =      7.7 : 1.0
                  seagal = True              neg : pos    =      7.4 : 1.0
                 balance = True              pos : neg    =      7.2 : 1.0
                 idiotic = True              neg : pos    =      7.1 : 1.0
              friendship = True              pos : neg    =      6.8 : 1.0
                   ideal = True              pos : neg    =      6.6 : 1.0
                   anger = True              pos : neg    =      6.0 : 1.0
               embarrass = True              neg : pos    =      6.0 : 1.0

Explaination:-  
-> if outstanding = True -> positive classification is more likely bcoz, pos:neg ratio = 12.6 : 1.0  
-> inane = True -> negative classification is more likely bcoz, neg:pos ratio = 5.5 : 1.0 